src/adapters/thinkerbell/mod.rs

```diff
@@ -0,0 +1,404 @@
+    timer: Arc<Mutex<timer::Timer>>
+}
+
+impl<T> ExecutableDevEnv for ThinkerbellExecutionEnv<T>
+    where T: API + Clone + Send + 'static {
+    // Don't bother stopping watches.
+    type WatchGuard = T::WatchGuard;
+    type API = T;
+
+    fn api(&self) -> Self::API {
+        self.adapter_manager.clone()
+    }
+
+    type TimerGuard = timer::Guard;
+    fn start_timer(&self, duration: Duration, sender: Box<ExtSender<()>>) -> Self::TimerGuard {
+        let timer = self.timer.lock().unwrap(); // FIXME: There's no way to report this error...
+        timer.schedule_with_delay(duration.into(), move || {
+            let _ = sender.send(());
+        })
+    }
+}
+
+/// Convert a ScriptManagerError into an API Error.
+/// We can't implement From<T> because ScriptManagerError is in a different crate.
+fn sm_error(e: ScriptManagerError) -> Error {
+    Error::InternalError(InternalError::GenericError(format!("{:?}", e)))
+}
+
+impl<T> Adapter for ThinkerbellAdapter<T> where T: AdapterManagerHandle + API + Clone + Send + 'static {
+    fn id(&self) -> Id<AdapterId> {
+        self.adapter_id.clone()
+    }
+
+    fn name(&self) -> &str {
+        ADAPTER_NAME
+    }
+
+    fn vendor(&self) -> &str {
+        ADAPTER_VENDOR
+    }
+
+    fn version(&self) -> &[u32;4] {
+        &ADAPTER_VERSION
+    }
+
+    fn fetch_values(&self, set: Vec<Id<Getter>>) -> ResultMap<Id<Getter>, Option<Value>, Error> {
+        set.iter().map(|id| {
+            let (tx, rx) = channel();
+            let _ = self.tx.send(ThinkAction::RespondToGetter(tx, id.clone()));
+            match rx.recv() {
+                Ok(result) => (id.clone(), result),
+                // If an error occurs, the channel died!
+                Err(recv_err) => (id.clone(), Err(Error::InternalError(
+                        InternalError::GenericError(format!("{:?}", recv_err))))),
+            }
+        }).collect()
+    }
+
+    fn send_values(&self, values: HashMap<Id<Setter>, Value>) -> ResultMap<Id<Setter>, (), Error> {
+        values.iter()
+            .map(|(id, value)| {
+                let (tx, rx) = channel();
+                let _ = self.tx.send(ThinkAction::RespondToSetter(tx, id.clone(), value.clone()));
+                match rx.recv() {
+                    Ok(result) => (id.clone(), result),
+                    // If an error occurs, the channel died!
+                    Err(recv_err) => (id.clone(), Err(Error::InternalError(
+                            InternalError::GenericError(format!("{:?}", recv_err))))),
+                }
+            })
+            .collect()
+    }
+
+    fn register_watch(&self, mut watch: Vec<(Id<Getter>, Option<Range>)>,
+        _: Box<ExtSender<WatchEvent>>) ->
+            ResultMap<Id<Getter>, Box<AdapterWatchGuard>, Error> {
+        watch.drain(..).map(|(id, _)| {
+            (id.clone(), Err(Error::GetterDoesNotSupportWatching(id)))
```

Comment:
Maybe in the future, we'll be able to watch whether a rule is enabled or disabled. Not now, though.